Implement two-stage sampling for AO/AOC analyze

darthunix opened the pull request:
To calculate statistics at first we have to retrieve in reservoir some representative sample from relation. PostgreSQL uses two stage sampling to avoid sequential scan of the whole relation:

* at first it samples blocks with algorithm S from Knuth
* for each block it samples tuples with algorithm Z from Vitter

To make it work every AM has to know how to estimate block number for any relation. These blocks should be of the fix size to make algorithm S from Knuth work correctly. Previously, AO/AOC relations didn't use two stage sampling and sequentially scanned the whole relation. The main reason for it was the fact that segment files are constructed from variable sized blocks (that are not suitable for algorithm S from Knuth).

Current implementation treats AO/AOC relations as a collection of segment files with a global sequence of rows among them:

* segno 1 with 100 rows - its rows are treated as global positions 0-99
* segno 2 with 200 rows - its rows are treated as global positions 100-299

This global sequence of rows is divided on some arbitrary number of "logical" analyze blocks fitting BlockNumber (uint32). These blocks are not interconnected with variable sized physical blocks of segment files - it is simply a range of global row positions. So block sampling by algorithm S from Knuth gives us the starting global position of a target block (the size of the block is fixed and known). We get a benefit of this schema when the number of blocks is more that 300 * statistics target.
So if we are lucky enough there would be several physical variable sized blocks between logical blocks and we can just read their headers without decompressing data. This is the main idea of speeding up AO/AOC analyze. After that we simply iterate among all rows in a logical analyze block and fill reservoir with samples.

To divide AO/AOC table into logical blocks we introduce a new GUC `gp_appendonly_analyze_block_size` controlling the default analyze block size. We divide AO/AOC table with logical blocks of this size. If the result number of blocks doesn't fit BlockNumber we replace it with the maximum possible number of blocks (`2^33-1`).

Current implementation is suboptimal in AOC projections at the beginning of analyze scan: AM API doesn't allow us to project only required columns for analyze scan and we have to scan all of them. It would be nice to improve AM cause a simple scan is initialised with such projection, but analyze scan is not.

I would also like to explain why we use fix sized logical blocks sampling with algorithm S from Knuth instead of weighted random sampling (WRS) algorithms (A-Chao or Efraimidis and Spirakis). At first I tried them but was confused with AOC sampling. Every column has its own variable sized blocks with different sizes and I didn't know what column to choose for WRS. The only thing in common among all columns segment files was the row count, so that was the reason to chose sampling based on fix sized logical blocks in a global sequence of rows in a table.

darthunix commented:
It seems to be a Travis internal error  and not interconnected with PR.

hlinnaka commented on src/backend/storage/buffer/bufmgr.c:
```diff
@@ -2962,17 +2964,43 @@ RelationGetNumberOfBlocksInFork(Relation relation, ForkNumber forkNum)
-				 * Therefore divide, rounding up.
-				 */
-				uint64		szbytes;
+				if (RelationIsAppendOptimized(relation))
```
This seems to assume that RelationGetNumberOfBlocks() is only used for ANALYZE. I don't think that's a good assumption, there are RelationGetNumberOfBlocks() calls in many places. We probably should rely less on it, in Postgres, now that we have table AMs, but I don't think changing its meaning is a good idea. In my mind, RelationGetNumberOfBlocks() should return the table's physical size on disk, divided by BLCKSZ.

darthunix replied:
Fixed in https://github.com/greenplum-db/gpdb/pull/11190/commits/14f9f4a2954612f1da4e2b764897be44fd46f421.

hlinnaka commented on src/backend/commands/analyze.c:
```diff
@@ -1388,32 +1388,7 @@ acquire_sample_rows(Relation onerel, int elevel,
+	totalblocks = RelationGetNumberOfBlocks(onerel);
 
 	/* Need a cutoff xmin for HeapTupleSatisfiesVacuum */
 	OldestXmin = GetOldestXmin(onerel, PROCARRAY_FLAGS_VACUUM);
```
I think a refactor of the table AM API's functions for ANALYZE would be in order. A suggestion off the top of my head:

Replace table_scan_analyze_next_block() and table_scan_analyze_next_tuple() with a single function: table_acquire_sample_rows(). The AM implementation of table_acquire_sample_rows() can use the BlockSampler funtions if it wants to, but if the AM is not block-oriented, it could do something else.

It would be good to submit a proposal for that to pgsql-hackers, because this affects any table AM implementations that are not block-oriented.

darthunix replied:
I can also suggest to refactor [table_beginscan_analyze()](https://github.com/greenplum-db/gpdb/blob/cb3ed1ea9491f42f8fb30b08a3985591dbdcc1b5/src/include/access/tableam.h#L888) to initialize projections for analyze in a manner [table_beginscan()](https://github.com/greenplum-db/gpdb/blob/cb3ed1ea9491f42f8fb30b08a3985591dbdcc1b5/src/include/access/tableam.h#L761) does for a regular scan (if it is possible). Should I modify AM for analyze in this PR before suggestion it to pgsql-hackers or start a discussion  there first?

hlinnaka replied:
Go ahead and start the discussion on pgsql-hackers. Or feel free to update this PR first, if you prefer, up to you.

BTW, I just bumped into this thread on hackers: https://www.postgresql.org/message-id/20201109180644.GJ16415%40tamriel.snowman.net. It adds a prefetching to ANALYZE. The reason I mention that is that that patch is also making the modularity violation of the table AM's analyze function worse, by adding more block-level code to the general analyze code, which IMHO belongs the table AM. So that's one more indication that the table AM API for ANALYZE needs to be refactored.

darthunix replied:
@hlinnaka I've opened a [thread](https://www.postgresql.org/message-id/flat/C7CFE16B-F192-4124-BEBB-7864285E0FF7%40arenadata.io) in psql-hackers about this problem.

hlinnaka commented on src/backend/access/aocs/aocsam_handler.c:
```diff
@@ -1222,12 +1240,206 @@ aoco_relation_copy_for_cluster(Relation OldHeap, Relation NewHeap,
+	return true;
+}
+
 static bool
```
Please add a comment explaining the algorithm here. You had that in the commit message, but it deserves to be in the code, too.

darthunix replied:
Fixed in https://github.com/greenplum-db/gpdb/pull/11190/commits/cb3ed1ea9491f42f8fb30b08a3985591dbdcc1b5.

bhuvnesh2703 commented:
@darthunix Hello Denis, I was testing the results of this PR and the numbers seems encouraging (46sec master vs 4sec this PR), i used the below setup
```
create table t3 (a int, b inet, c inet) with (appendonly=true, orientation=column, compresstype=zlib, compresslevel=3);insert into t3 select i, (select ((i%255)::text \|\| $.$ \|\| (i%255)::text \|\| $.$ \|\| (i%255)::text \|\| $.$ \|\| (i%255)::text))::inet, (select ((i%255)::text \|\| $.$ \|\| (i%255)::text \|\| $.$ \|\| (i%255)::text \|\| $.$ \|\| (i%255)::text))::inet from generate_series(1,10000000)i;
insert into t3 select * from t3;
insert into t3 select * from t3;
insert into t3 select * from t3;
insert into t3 select * from t3;
insert into t3 select * from t3;
insert into t3 select * from t3;
insert into t3 select * from t3;
select count(*) from t3;   
count
------------ 
1280000000(1 row)
--
```
I read the upstream thread, and it appears that it required a redesign, and the last comment was on 2021-10-01.
Wanted to cycle back on to it and understand the current status or the plan you have?
Let us know how we can help to get it to closure.

darthunix commented:
@bhuvnesh2703 current PR is abandoned because no one was interested in it.
- Arenadata uses 6X branch and doesn't want to modify the stable code of the analyze command (i.e. buy performance for bugs). So, it was decided to experiment with analyze command in the master branch of GP. That was the reason to open current PR to 7X.
- Pivotal wanted to keep 7X as much close to vanilla PG as possible, but the current PR modifies analyze code inherited from PG. So @hlinnaka suggested to refactor analyze command in PG first and then pulling the result commit to GP. At first, I've opened a conversation and even suggested a patch to CF... but later understood, that PG supports only a single storage manager with the blocks of a fixed size and a shared buffer (AOCO uses an additional storage manager with direct fsync without buffering). So my patch for variable sized blocks was useless for PG until it doesn't support other storage managers. I removed it from CF and closed the discussion.

Then Heikki moved to Zenith and PR became abandoned. I've switched to other tasks (now they are not even related to GP). Are you interested in the current PR or have any ideas on how to improve it?

bhuvnesh2703 commented:
@darthunix Thanks for giving the context. The performance results looks exciting, so we are definitely interested to understand it more closely. We are working to improve the performance for analyze on 7x and 6x, and for AOCO table specifically (which i tried) it seems to help. So, will be spending some time on understanding it and taking it further. Feel free to add any inputs.

soumyadeep2007 commented:
@kainwen since you and team are looking at this, we should also knock this related FIXME out as a separate commit in this PR:
```c
/*
 * GPDB_12_MERGE_FIXME: For AO/AOCO tables, blocknumber does not have a
 * meaning and is not set. The current implementation of analyze makes
 * assumptions about the file layout which do not hold for these two cases.
 * The compare function should maintain the row order as consrtucted, hence
 * return 0;
```

haolinw commented:
> @kainwen since you and team are looking at this, we should also knock this related FIXME out as a separate commit in this PR:
> 
> ```c
> /*
>  * GPDB_12_MERGE_FIXME: For AO/AOCO tables, blocknumber does not have a
>  * meaning and is not set. The current implementation of analyze makes
>  * assumptions about the file layout which do not hold for these two cases.
>  * The compare function should maintain the row order as consrtucted, hence
>  * return 0;
> ```

Sure, will take it forward.

haolinw commented:
Opened PR https://github.com/greenplum-db/gpdb/pull/14416 to proceed this work.

ashwinstar commented:
Closing this as we are continuing to take the work forward via PR #14416 now. Thank you @darthunix so much for contribution. Would be great if can help review/comment on the newer PR from Haolin.
